### PR TITLE
ENH: Added sanity check to printoptions

### PR DESCRIPTION
**Pull request by Mitchell-Faas**
See issue #18254

### CODE REVIEW

**Comment by rgommers** [comment]
Restarting CI, something weird happened.

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -78,13 +78,20 @@ def _make_options_dict(precision=None, threshold=None, edgeitems=None,
...
         # forbid the bad threshold arg suggested by stack overflow, gh-12351
         if not isinstance(threshold, numbers.Number):
             raise TypeError("threshold must be numeric")
         if np.isnan(threshold):
             raise ValueError("threshold must be non-NAN, try "
                              "sys.maxsize for untruncated representation")
+
+    if precision is not None:
+        # forbid the bad precision arg as suggested by issue #18254
+        if not isinstance(precision, int):
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:92]: Perhaps better as `precision=operator.index(precision)`, else this blows up if users pass an `np.int32` or something else that's harmless

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -78,13 +79,22 @@ def _make_options_dict(precision=None, threshold=None, edgeitems=None,
...
         if np.isnan(threshold):
             raise ValueError("threshold must be non-NAN, try "
                              "sys.maxsize for untruncated representation")
+
+    if precision is not None:
+        # forbid the bad precision arg as suggested by issue #18254
+        try:
+            options['precision'] = index(precision)
+        except TypeError:
+            raise TypeError('precision must be an integer')
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:96]:
```suggestion
        except TypeError as e:
            raise TypeError('precision must be an integer') from e
```

File: numpy/core/arrayprint.py [hunk 3]
```diff
@@ -41,6 +41,7 @@
 from .numerictypes import (longlong, intc, int_, float_, complex_, bool_,
                            flexible)
 from .overrides import array_function_dispatch, set_module
+from operator import index
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:44]:
```suggestion
import operator
```

File: numpy/core/arrayprint.py [hunk 4]
```diff
@@ -78,13 +79,22 @@ def _make_options_dict(precision=None, threshold=None, edgeitems=None,
...
         if not isinstance(threshold, numbers.Number):
             raise TypeError("threshold must be numeric")
         if np.isnan(threshold):
             raise ValueError("threshold must be non-NAN, try "
                              "sys.maxsize for untruncated representation")
+
+    if precision is not None:
+        # forbid the bad precision arg as suggested by issue #18254
+        try:
+            options['precision'] = index(precision)
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:94]:
```suggestion
            options['precision'] = operator.index(precision)
```

**Review by eric-wieser — COMMENTED** [review]
I think we normally don't do `from operator import *`, as those names are very likely to overlap with other names. If nothing else, it makes `operator.index` something which can be searched for.

File: numpy/core/arrayprint.py (hunk 3 above)
Comment (Mitchell-Faas) [reply numpy/core/arrayprint.py:44]: Is it not beneficial to keep the amount of imported material to a minimum, especially looking at potential performance considerations?

Comment (eric-wieser) [reply numpy/core/arrayprint.py:44]: That's not how importing works. `from module import x` is just `import module; x = getattr(module, 'x'); del module`

**Comment by BvB93** [comment]
Can you update the annotations in [arrayprint.pyi](https://github.com/numpy/numpy/blob/master/numpy/core/arrayprint.pyi) as well?

In total two changes will be necessary:

1. Import the `SupportsIndex` protocol right [here](https://github.com/numpy/numpy/blob/f55f5d45518c2f83271330041b982e55dcd6bbf9/numpy/core/arrayprint.pyi#L26-L29):
``` python
if sys.version_info > (3, 8):
    from typing import SupportsIndex
else:
    from typing_extensions import SupportsIndex
```

2. Replace `int` with `SupportsIndex` (for the precision parameter) in the following functions:
* [`set_printoptions`](https://github.com/numpy/numpy/blob/f55f5d45518c2f83271330041b982e55dcd6bbf9/numpy/core/arrayprint.pyi#L65)
* [`array2string`](https://github.com/numpy/numpy/blob/f55f5d45518c2f83271330041b982e55dcd6bbf9/numpy/core/arrayprint.pyi#L82)
* [`array_repr`](https://github.com/numpy/numpy/blob/f55f5d45518c2f83271330041b982e55dcd6bbf9/numpy/core/arrayprint.pyi#L120)
* [`array_str`](https://github.com/numpy/numpy/blob/f55f5d45518c2f83271330041b982e55dcd6bbf9/numpy/core/arrayprint.pyi#L126)
* [`printoptions`](https://github.com/numpy/numpy/blob/f55f5d45518c2f83271330041b982e55dcd6bbf9/numpy/core/arrayprint.pyi#L133)

**Review by eric-wieser — APPROVED** [review]

**Review by BvB93 — APPROVED** [review]
The annotation update is looking good as well.

**Comment by charris** [comment]
Thanks @Mitchell-Faas .

**Pull request merged**